lib/srv/mcp/stdio_test.go

```diff
@@ -0,0 +1,359 @@
+	case <-handlerDoneCh:
+	}
+	event := emitter.LastEvent()
+	_, ok := event.(*apievents.MCPSessionEnd)
+	require.True(t, ok)
+}
+
+func makeMockMCPServerRunner(context.Context, *sessionHandler) (stdioServerRunner, io.ReadCloser, io.Writer, error) {
+	serverStdin, writeToServer := io.Pipe()
+	readFromServer, serverStdout := io.Pipe()
+	return &mockStdioServerRunner{
+		serverStdin:  serverStdin,
+		serverStdout: serverStdout,
+		pipeClosers: []io.Closer{
+			serverStdin, writeToServer,
+			readFromServer, serverStdout,
+		},
+	}, readFromServer, writeToServer, nil
+}
+
+type mockStdioServerRunner struct {
+	serverStdin  io.Reader
+	serverStdout io.Writer
+	pipeClosers  []io.Closer
+}
+
+func (s *mockStdioServerRunner) run(ctx context.Context) error {
+	slog.DebugContext(ctx, "running mock stdio server")
+	server := mcpserver.NewMCPServer("test-server", "1.0.0")
+	server.AddTool(mcp.Tool{
+		Name: "hello-server",
+	}, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+		return &mcp.CallToolResult{
+			Content: []mcp.Content{mcp.NewTextContent("hello client")},
+		}, nil
+	})
+	err := mcpserver.NewStdioServer(server).Listen(ctx, s.serverStdin, s.serverStdout)
+	if err != nil && !mcputils.IsOKCloseError(err) {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (s *mockStdioServerRunner) close() {
+	for _, pipeCloser := range s.pipeClosers {
+		pipeCloser.Close()
+	}
+}
+
+func makeStdioClient(t *testing.T, clientSourceConn net.Conn) *mcpclient.Client {
+	t.Helper()
+	stdioClientTransport := mcpclienttransport.NewIO(clientSourceConn, clientSourceConn, io.NopCloser(bytes.NewReader(nil)))
+	stdioClient := mcpclient.NewClient(stdioClientTransport)
+	t.Cleanup(func() {
+		stdioClient.Close()
+	})
+	require.NoError(t, stdioClient.Start(t.Context()))
+	return stdioClient
+}
+
+func initializeStdioClient(ctx context.Context, client *mcpclient.Client) (*mcp.InitializeResult, error) {
+	initReq := mcp.InitializeRequest{}
+	initReq.Params.ProtocolVersion = mcp.LATEST_PROTOCOL_VERSION
+	initReq.Params.ClientInfo = mcp.Implementation{
+		Name:    "test-client",
+		Version: "1.0.0",
+	}
+	resp, err := client.Initialize(ctx, initReq)
+	return resp, trace.Wrap(err)
+}
+
+// TestHandleSession_execMCPServer tests real server handler for stdio-based MCP
+// server but requires docker installed locally. It will run the "everything"
+// MCP server and "alpine".
+//
+// TELEPORT_MCP_TEST_DOCKER_WITH_HOST_USER needs to be set with a username to
+// run as for the docker command. To test running with ambient credential vs a
+// different user, run the test once with your current user as host user then
+// run one more time with sudo.
+func TestHandleSession_execMCPServer(t *testing.T) {
+	runAsHostUser := os.Getenv("TELEPORT_MCP_TEST_DOCKER_WITH_HOST_USER")
+	if runAsHostUser == "" {
+		t.Skip("This test requires docker and set TELEPORT_MCP_TEST_DOCKER_WITH_HOST_USER=<host_user_name> in your environment")
+	}
```

Comment:
Neat; I think there is still a way to make it runnable on CI, right?

Reply:
ya, probably can set up real containers for e2e test. too lazy to do that for now.